proposals/accepted/051-ghc-base-libraries.rst

```diff
@@ -0,0 +1,330 @@
+
+3. We should use this pattern for ``base``.
+
+4. The Core Libraries Committee curates the API of ``base`` (here is `the charter <https://github.com/haskell/core-libraries-committee#base-package>`__), including:
+
+   - Types, specifically including what instances are exposed
+
+   - Semantics (including strictness)
+
+   - Performance
+
+   - Semantic changes to documentation (the charter says *"Documentation changes normally fall under GHC developers purview, except significant ones (e.g., adding or changing type class laws)."*)
+
+   Beyond that, it has no interest in the implementation details (e.g. alpha renaming, moving things between modules, comments).
+
+5. In curating the ``base`` API, it is immaterial where code lives.
+   For example, if a change to ``ghc-prim`` changes the ``base`` API (as defined above) the GHC developers must consult the CLC.
+   The fact that the change isn't physically part of the ``base`` package is immaterial.
+   (Incidentally, ``base`` and ``ghc-prim`` are both part of the same GitHub repository, which also includes GHC's source code.)
+
+Proposal
+========
+
+We propose to divide ``base`` into three packages:
+
+- ``ghc-internals``: exposes aspects of GHC's internals that may be of interest to "hard-core" developers interested in maximum performance (see `Nikita's blog post <https://nikita-volkov.github.io/internal-convention-is-a-mistake/>`__).
+  The API of ``ghc-internals`` is fully under the control of the GHC team, and of no direct interest to the CLC — only its effects on the API of base.
+
+- ``base``: as now, whose API is curated by CLC.
+  Depends on ``ghc-internals``, and hence on ``ghc-bignum`` and ``ghc-prim``.
+
+- ``ghc-experimental``, initially empty, depends on ``base`` and on ``ghc-internals``.
+  Functions and data types here are intended to have their ultimate home in base, but while they are settling down they are subject to much weaker stability guarantees.
+  Generally, new functions and types introduced in GHC Proposals would start their life here.
+  Example: new type families and type constructors for tuples, `GHC Proposal #475 <https://github.com/ghc-proposals/ghc-proposals/pull/475>`__.
+
+  Another example: future APIs to access RTS statistics, which are fairly stable and user-exposed, but which are (by design) coupled closely to GHC's runtime and hence may change.
+
+  As its name suggests, the API of ``ghc-experimental`` is curated by the GHC team, although the CLC is willing to offer (non-binding) opinions, if consulted.
+
+All three packages conform rigorously to the PVP.
+(But see Section 5.3)
+
+Some observations about this structure:
+
+- We should use all possible social and technical means to discourage people from depending directly on ``ghc-internals``, because if such dependencies become frequent and ossified, it will lead to future pain when the API changes.
+  Saying "we told you not to rely on it" may be true, but won't lessen that pain.
```

Comment:
I also don't feel like this part of the proposal is well-motivated - specifically why `ghc-internals` should be treated separately from a potential `text-internals` or `bytestring-internals`. 

From my perspective, writing `build-depends: ghc-internals` is just as prone to pain as any other unbounded dependency. And `build-depends: ghc-internals == 9.6.4.*` should be perfectly safe, unless PVP is violated. But the *point* of this proposal is that we can make "breaking changes" to `ghc-internals` as a major version bump without incurring a major version bump for `base` - in other words, so we *can* be PVP compliant.

Reply:
> I also don't feel like this part of the proposal is well-motivated - specifically why ghc-internals should be treated separately from a potential text-internals or bytestring-internals.

It's really a response to a legitimate CLC concern that if it's too easy to depend on `ghc-internals` then that's what people will do.  And if that happens we might grow an ecosystem of libraries many of whch depend, perhaps by accident, on `ghc-internals`.  So every new release of GHC would force updates to all of those libraries.  You could say "if you depend on `ghc-internals` then that's what you get, but the pain is real and not mitigated by saying "I told you so".  To put it another way, it risks entirely bypassing the stability efforts of the CLC.

So that is, I think, the reason for this text. You are right that all the same issues apply to any `-internals` package, and one might wonder about mechanisms to discourage depending on them. But to avoid scope creep in this document we just stuck to `ghc-internals`.

Reply:
Discouragement is well motivated. Prohibition isn't.

We have to educate users more generally about the pattern we're trying to introduce here and how it relates to PVP, API stability, etc.

I could imagine that we try to summarize the result of these efforts for end users and library maintainers in a more distilled way and explain the pattern there in more depth.

I don't think we have to be particularly scared about ecosystem degrading.

1. Hackage has a reverse deps feature and we can monitor packages depending on ghc-internals easily.
2. There are many eyes on popular packages like aeson, servant, conduit etc. and I'm sure someone will scold any attempt at introducing a ghc-internals dependency
3. Even if a package accidentally does so, it's not irreversible. We can scold the maintainer and ask why they saw the need to do so.
4. "Pick your libraries carefully" already applies. Well educated maintainers who care about the ecosystem will play nicely.

Reply:
@simonpj I think that makes good sense. I want GHC to be able to innovate freely and iterate quickly, and I wouldn't want CLC to start encroaching on the `ghc-internals` package for this reason. Where I think I'd draw the line is at describing the problem, noting it is a known ecosystem-wide problem that deserves a good solution, and somehow put a reminder on the package to *use* that solution when it is implemented.

There are a few perspectives here, and I think it'd be good to be really clear about which perspective a bullet is trying to satisfy. This one appears to be trying to satisfy both "industrial users want fewer breaking changes when upgrading major versions of GHC" and "CLC wants to be able to promote stability for the ecosystem" - which are overlapping but subtly different concerns.

As an industry user and OSS contributor, the primary pain point of upgrading GHC is that there are breaking changes that require a large chunk of the ecosystem to be touched. `base` is usually a relatively small part of this pain, with `ghc-prim` and `template-haskell` causing far more issues. I don't expect that `ghc-internals` is going to meaningfully impact the work that gets done here.

If `ghc-internals` is banned by `cabal-install`, then I don't see how it is different from un-exposed modules in `base`. If `ghc-internals` is banned on Hackage, then industrial users will need to vendor code in order to share libraries that depend on it - which *increases* the pain of a major upgrade by requiring *applications* to depend on `ghc-internals` directly, rather than sharing a library which can abstract over multiple versions of `ghc-internals` via CPP or other techniques. If there are warnings, they will be ignored for good and bad reasons, and maintainers will have another package to consider when doing ecosystem upgrades, but the overall workload doesn't change much.

As a CLC member, `ghc-internals` makes my life easier by having fewer changes to `base`, and changes to `base` that can be backed up by real ecosystem use in `ghc-internals`. But if using `ghc-internals` is heavily discouraged, then it's difficult to identify whether or not something in that package is satisfying the needs of the ecosystem.

Summarizing,

- I don't think we should use strong wording like "all possible social and technical means." 
- I'm not even sure we need to discourage beyond calling the package `ghc-internals` and writing docs that it'll be a major version bump with breaking changes every GHC release.
- I *definitely* think that proposing specific technical means for discouraging folks is out-of-scope here.

Reply:
The more I think about it, the more I am convinced that "all possible social and technical means" is too strong. It is obviously not literal. We are not going to hire a crack commando unit who survive as soldiers of fortune to go to your house if you depend on it.

Better language might be "develop both social and technical mechanisms to discourage..."